tplink_smartplug.java

```diff
@@ -0,0 +1,180 @@
+ *         For use with TP-Link HS-100 or HS-110<br/>
+ */
+public class tplink_smartplug
+{
+
+   /**
+    * Port that the Smart Plug listens on
+    */
+   private final static int PORT = 9999;
+
+   /**
+    * Key that the XOR Cipher starts with
+    */
+   private final static int KEY = 171;
+
+   /**
+    * Main method used to send and receive data from the Smart Plug
+    * 
+    * @param args
+    *           Argument 1: IP Address<br/>
+    *           Argument 2: Command to Send<br/>
+    */
+   public static void main(String[] args)
+   {
+      Socket socket = null;
+      try
+      {
+         socket = new Socket(args[0], PORT);
+         System.out
+               .println(socket.isConnected() ? "Connected to Device" : "Device Connection Failed");
+
+         // No point wasting CPU cycles if the socket isin't connected
+         if (!socket.isConnected())
+         {
+            sendCommand(args[1], socket);
+            readResponse(socket);
+         }
+
+      }
+      catch (IOException e)
+      {
+         e.printStackTrace();
+      }
+      finally
+      {
+         if (socket != null)
+         {
+            try
+            {
+               System.out.println("Closing");
+               socket.close();
+            }
+            catch (IOException e)
+            {
+               e.printStackTrace();
+            }
+         }
+      }
+
+      System.out.println("Closed");
+   }
+
+   /**
+    * Send a command to the smart plug
+    * 
+    * @param command
+    *           Command to send to the Smart Plug
+    * @param socket
+    *           Socket to send the command on
+    * @throws IOException
+    *            Failure of socket communications
+    */
+   private static void sendCommand(String command, Socket socket) throws IOException
+   {
+      System.out.println("Sending Command: " + command);
+
+      DataOutputStream writer = new DataOutputStream(socket.getOutputStream());
+      writer.write(encrypt(command));
+   }
+
+   /**
+    * Read a response from the Smart Plug
+    * 
+    * @param socket
+    *           Socket to send the command on
+    * @throws IOException
+    *            Failure of socket communications
+    */
+   private static void readResponse(Socket socket) throws IOException
+   {
+      DataInputStream reader = new DataInputStream(socket.getInputStream());
+      ArrayList<Byte> result = new ArrayList<Byte>();
+      while (reader.available() > 0)
+      {
+         result.add(reader.readByte());
+      }
+
+      System.out.println(decrypt(result.toArray(new Byte[result.size()])));
+   }
+
+   /**
+    * The smart plug communicates with the client via XOR Ciphered JSON
+    * 
+    * This encrypts the JSON to the format the plug expects
+    * 
+    * @param stringToEncrypt
+    *           Command to encrypt
+    * @return byte array ready to be sent to the device
+    */
+   static byte[] encrypt(String stringToEncrypt)
+   {
+      int index = 4; // First 4 bytes are 0x00
```

Comment:
First 3 bytes are 0x00, byte 4 is now length of string

Reply:
It would seem reasonable that _all four bytes_ encode the length of the string, as a big-endian, unsigned integer. This is what the current Python code implements:
https://github.com/softScheck/tplink-smartplug/blob/395c3521000f78bacd0d0cb32b6c6faf825f6c1a/tplink_smartplug.py#L56